### DISCUSSION TITLE
in update case setting nodes and edges state from backend but rfInstance is null

### DISCUSSION THREAD

**Discussion by shubh-am8** [xyflow/xyflow#2089] [Q&A]
first of all thanks a lot for this wonderful library ,some times nodes and edges state is bit tricky to handel.
in my case i have the nodes and edges data in backend on rendring the page i had setting the node,edges state from fetched data and the flow rendered smoothly but when i want to add new node or want to make changes in the existing node then in my handelFunction 
`const handelOnDeleteBtnChoice = (choiceId, nodeId) => {
    let instance = rfInstance.toObject()
    const selectedNode = instance.nodes.filter(function (node) {
      return node.id == nodeId;
    }).pop();
    let filteredChoices = selectedNode.data.choices.filter(function (c) {
      return c.id != choiceId;
    });
    selectedNode.data.choices = filteredChoices
    setNodes((nodes) => applyNodeChanges([selectedNode], nodes));
    console.log(nodes)
  };`
  
  but in this case rfinstsnse is null;
even the state isn't mutated so i have no idea how to achieve my goal pls help !

**Comment by bcakmakoglu** [comment]
Off the top of my hat I would guess that you might have forgotten to wrap your Flow with the `<ReactFlowProvider>`, which is required according to the [docs](https://reactflow.dev/docs/guides/uncontrolled-flow/#updating-nodes-and-edges), otherwise using `const reactFlowInstance = useReactFlow();` wouldn't work. 

There's also another issue in your code.
```jsx
setNodes((nodes) => applyNodeChanges([selectedNode], nodes)); console.log(nodes) };
```

`applyNodeChanges` accepts an Array of `NodeChange` not an Array of `Node` which have been changed.
You can find the types [here](https://github.com/wbkd/react-flow/blob/main/src/types/changes.ts).

This is how your code should look, if I'm not mistaken.
```jsx
function Flow() {
  const reactFlowInstance = useReactFlow();

  const onClick = (choiceId, nodeId) => {
   const selectedNode = reactFlowInstance.getNode(nodeId)

   selectedNode.data.choices = selectedNode.data.choices.filter((c) => { 
       return c.id !== choiceId; 
    })

    reactFlowInstance.setNodes((nodes) => {
      return nodes.filter((n) => n.id !== nodeId).concat(selectedNode);
    });
  };

  ... the rest of your jsx
}

export default function () {
  return (
    <ReactFlowProvider>
      <Flow />
    </ReactFlowProvider>
  );
}
```

You can find an example in this [sandbox](https://codesandbox.io/s/sandpack-project-forked-y14j14).

I hope this helps 😄
